### ISSUE TITLE
Rule validation functionality

### ISSUE THREAD

**Issue by DullReferenceException** [godaddy/node-redirect-rules#3]
Suggestion from @JosephJNK is to be able to detect issues with a set of redirect rules, for example:
1. Degenerate regex that would cause a DOS
2. Redirect loops
3. Unnecessary redirect hops, like `[A -> B, B -> C]` instead of `[A -> C, B -> C]`

**Comment by apoco** [comment]
Should also emit warnings for redundancy in rules since the rule syntax is permissive, allowing multiple ways to specify the same things.

**Comment by asilvas** [comment]
IMO this validation step should be a helper that can be done at the time of saving the rule. To avoid constant revalidation on every single request.

**Comment by DullReferenceException** [comment]
Yes, was thinking there'd be an API method for validating the rules, not that it'd be done at runtime. So if a caller was, say, storing the redirect rules in a database, it could block storing those rules if the validation didn't pass.

**Comment by DullReferenceException** [comment]
Example:

```
var redirectRules = require('redirect-rules');

var candidateRules = getRulesFromSomewhere();
var validationResult = redirectRules.validate(candidateRules);
if (validationResult.errors.length) {
  sendCallerValidationMessageBasedOn(validationResult);
}
```

**Comment by asilvas** [comment]
+1
